File: sol/974/974.cpp
#include <unordered_map>

struct Solution {
    int subarraysDivByK(vector<int>& nums, int k) {
        unordered_map<int, int> d;
        d[0] = 1;
        int prefix = 0, res = 0;
        for (int &x : nums){
            prefix = ((prefix + x) % k + k) % k;
            res += d[prefix]++;
        }
        return res;
    }
};

File: sol/974/974.py
class Solution:
    def subarraysDivByK(self, nums: List[int], k: int) -> int:
        d = {0: 1}
        prefix = 0
        for x in nums:
            prefix = (prefix + x) % k
            d[prefix] = d.get(prefix, 0) + 1
        res = 0
        for v in d.values():
            res += v * (v - 1) // 2
        return res
